Feature/chung step1

kochungcheon opened the pull request:
[exception]
- 에러 코드를 어떻게 하면 잘 전달해줄지 고민입니다

[Metadata]
- 파일의 메타데이터 엔티티 입니다
- 검증 메서드(validate)와 실제 저장 경로 저장 메서드(storedName)가 있습니다

[FileReader, FileWriter, FileDataAccessService]
- 의존성 역전 원칙 : 비지니스 로직이 서비스 로직에 의존하지 않게 됩니다
- 단일 책임의 원칙 : 읽기와 쓰기를 담당하는 인터페이스로 분리했습니다

[dto]
- dto를 만드는 기준이 고민입니다 인자가 몇개 부터 만들지..

[FileController]
- 클라이언트에게 성공 실패만 넘겨주면 된다는 생각에 204를 넘깁니다

kochungcheon committed:
feat: step1 구현

[exception]
- 에러 코드를 어떻게 하면 잘 전달해줄지 고민입니다

[Metadata]
- 파일의 메타데이터 엔티티 입니다
- 검증 메서드(validate)와 실제 저장 경로 저장 메서드(storedName)가 있습니다

[FileReader, FileWriter, FileDataAccessService]
- 의존성 역전 원칙 : 비지니스 로직이 서비스 로직에 의존하지 않게 됩니다
- 단일 책임의 원칙 : 읽기와 쓰기를 담당하는 인터페이스로 분리했습니다

[dto]
- dto를 만드는 기준이 고민입니다 인자가 몇개 부터 만들지..

[FileController]
- 클라이언트에게 성공 실패만 넘겨주면 된다는 생각에 204를 넘깁니다

kochungcheon committed:
refactor: 코드스멜 개선

- 안쓰는 import 파일 제거했습니다

kochungcheon committed:
refactor: Replace code style XML

- Replaced the XML file for code styling.

kochungcheon committed:
refactor: 코딩 컨벤션 적용

네이버 코딩 컨벤션 적용

Commit:
chore: update checkstyle version

Commit:
Merge branch 'main' into feature/chung_step1

VSFe committed:
chore: remove unsolved test

kochungcheon committed:
chore : 코딩 컨벤션 맞춰 수정

- 메서드 내 띄워쓰기

kochungcheon committed:
refactor: 테스트 코드 추가

- FileServiceTest / MetadataTest / ServiceExceptionTest 추가

kochungcheon committed:
refactor: 코드스멜 제거

- public 제거

kochungcheon committed:
refactor: 테스트 코드 추가

- 업로드 실패 테스트 코드 추가

kochungcheon committed:
refactor: 테스트 yml 제거

- 미사용으로 제거

kochungcheon committed:
refactor: 테스트 코드 추가

- 메타 데이터 없을 시 테스트 코드 추가

kochungcheon committed:
chore: 테스트 명 수정

- 업로드 테스트 -> 업로드 IO 테스트

VSFe committed:
test: 어쩌다 채운 코드

VSFe commented on gradlew.bat:
```diff
@@ -2,7 +2,7 @@
 @rem
 @rem Licensed under the Apache License, Version 2.0 (the "License");
-@rem you may not use this file except in compliance with the License.
+@rem you may not use this metadata except in compliance with the License.
```
일괄적으로 변경하다 발생한 대참사

VSFe commented on src/main/java/com/c4cometrue/mystorage/dto/FileDeleteRequest.java:
```diff
@@ -0,0 +1,11 @@
+
+import jakarta.validation.constraints.NotNull;
+
+public record FileDeleteRequest(
```
```java
public record FileDeleteRequest(
	@NotNull(message = "파일 id는 null 될 수 없습니다") Long fileId,
	@NotNull(message = "사용자 id는 null 될 수 없습니다") Long userId
) {
	public static FileDeleteRequest of(Long fileId, Long userId) {
		return new FileDeleteRequest(fileId, userId);
	}
}
```

VSFe replied:
NotNull -> 그냥 long으로 쓰는게 낫습니다.

kochungcheon replied:
확실히 널이 될 수 없는 값은 원시값으로 하는 게 좋은 거 같습니다!

VSFe commented on src/main/java/com/c4cometrue/mystorage/file/FileController.java:
```diff
@@ -0,0 +1,38 @@
+public class FileController {
+	private final FileService fileService;
+
+	@PostMapping("/files/upload")
```
"/files"는 묶어서 위로 올려보내는게 낫겠네요.

kochungcheon replied:
넵 공통 부분은 위로 올렸습니다!

VSFe commented on src/main/java/com/c4cometrue/mystorage/file/FileController.java:
```diff
@@ -0,0 +1,38 @@
+	@PostMapping("/files/upload")
+	public ResponseEntity<Void> uploadFile(FileUploadRequest fileUploadRequest) {
+		fileService.uploadFile(fileUploadRequest);
+		return ResponseEntity.noContent().build();
```
`@ResponseStatus(HttpStatus.NO_CONTENT)` 로 대체하고, 그냥 void를 반환하게 해 주세요.

kochungcheon replied:
어노테이션으로 상태코드 처리 할 수 있다는 것을 배웠습니다. 이에 맞춰 수정했습니다

VSFe commented on src/main/java/com/c4cometrue/mystorage/file/FileController.java:
```diff
@@ -0,0 +1,38 @@
+		return ResponseEntity.noContent().build();
+	}
+
+	@DeleteMapping("/files/delete")
```
일반적으로, 동작에 대한 정의를 HTTP Method로 설정하고, 주소에서는 동사적 행위를 최대한 생략합니다.

- 파일 업로드: POST /files
- 파일 삭제: DELETE /files
- 파일 다운로드: GET /files

와 같은 방식으로 수정해주세요.

kochungcheon replied:
RESTFUL 하게 HTTP Method를 정의하겠습니다

VSFe commented on src/main/java/com/c4cometrue/mystorage/file/FileDataAccessService.java:
```diff
@@ -0,0 +1,28 @@
+		);
+	}
+
+	public void persist(Metadata metadata) {
```
파일 업로드 과정에서, 문제가 되는 경우가 있지 않을까요?

- 사용자가 똑같은 이름의 파일을 연속해서 올리는 경우 (올라가선 안 됨.)
- (굳이 지금 당장은 구현할 필요가 없겠지만) 사용자가 사용해서는 안 되는 금칙어를 파일 이름 안에 삽입하는 경우

이런 것들을 persist 메서드 안에서 구현할 필요가 있어보입니다.

kochungcheon replied:
똑같은 이름을 가지는 파일을 체크하는 기능을 만들어서 처리했습니다.

VSFe commented on src/main/java/com/c4cometrue/mystorage/file/FileDataAccessService.java:
```diff
@@ -0,0 +1,28 @@
+	private final FileRepository fileRepository;
+
+	public void deleteBy(Long fileId) {
+		fileRepository.deleteById(fileId);
```
**파일이 존재하지 않은 상태에서 삭제를 시도한다면,** 이거는

- 정상적으로 처리가 되어야 한다.
- 아니다, NOT_EXIST_FILE 예외를 반환해야 한다.

어떤 선택을 할지 잘 생각해 보세요. (답은 없지만, 본인의 선택에 대한 근거가 중요하겠죠?)

VSFe replied:
혹시나 파일 검증을 할거라면, exist 메서드를 따로 만드는게 낫습니다.

(findBy를 호출해서 notNull 여부를 봐도 되겠지만, 굳이 데이터를 불러올 필요는 없으니까요.)

kochungcheon replied:
exist 메서드를 만들어 처리했습니다. 그리고 파일이 존재하지 않는 상태에서 삭제를 시도하면 에러를 반환했습니다. 파일을 삭제해야 정상처리 되는 게 자연스럽다 생각해서 입니다

VSFe commented on src/main/java/com/c4cometrue/mystorage/file/FileService.java:
```diff
@@ -0,0 +1,93 @@
+	private String storagePath;
+
+	@Value("${file.buffer}")
+	private Integer bufferSize;
```
어차피 NotNull이면 int로 해도 되지 않을까요?

kochungcheon replied:
널 값이 안들어 오기 때문에 원시타입으로 바꾸겠습니다

VSFe commented on src/main/java/com/c4cometrue/mystorage/file/FileService.java:
```diff
@@ -0,0 +1,93 @@
+	private Integer bufferSize;
+
+	@Transactional
+	public void uploadFile(FileUploadRequest request) {
```
개인적으로, Service 레이어에서 데이터를 주고 받을 때는 DTO 형식의 값이 전달되지 않도록 하는게 좋다고 생각해요.

DTO는 Controller 단에서 주고받는 데이터이기에, 레이어간 독립을 보장하기 위해선 다소 불편할 수 있어도 Unpack해서 전달하는게 맞을 거예요.

kochungcheon replied:
넵 알겠습니다! 서비스 레이어에는 Unpack 해서 전달하겠습니다. DTO를 사용할 때도 그 의미를 파악하고 적절한 곳에 사용해야 된다고 생각합니다

VSFe commented on src/test/java/com/c4cometrue/mystorage/file/FileDataAccessServiceTest.java:
```diff
@@ -0,0 +1,66 @@
+	FileRepository fileRepository;
+
+	@BeforeEach
+	public void setUp() {
```
`@ExtendWith(MockitoExtension.class)` 로 통일하죠.

kochungcheon replied:
코드 통일을 하겠습니다!

VSFe commented on src/main/java/com/c4cometrue/mystorage/file/FileService.java:
```diff
@@ -0,0 +1,93 @@
+		Long userId = request.userId();
+		String userPath = request.userPath();
+
+		Metadata metadata = fileDataAccessService.findBy(fileId);
```
그냥 findBy 할 때 fileId, userId를 모두 받아서 처리하게 하는게 더 자연스러울 것 같습니다.

(metadata.validate 제거)

kochungcheon replied:
확실히 이렇게 하는 게 더 자연스럽네요

VSFe commented on src/main/java/com/c4cometrue/mystorage/file/FileService.java:
```diff
@@ -0,0 +1,93 @@
+		Metadata metadata = Metadata.of(file.getOriginalFilename(), storedFileName, path.toString(), userId);
+		fileDataAccessService.persist(metadata);
+
+		try (InputStream is = file.getInputStream(); OutputStream os = Files.newOutputStream(path)) {
```
(리뷰가 아니고, 생각해보라는 의미에서 작성합니다.)

- (물리적 파일 업로드) -> (DB 파일 메타데이터 등록)
- (DB 파일 메타데이터 등록) -> (물리적 파일 업로드)

뭐가 더 자연스러울 것 같나요?

kochungcheon replied:
(물리적 파일 업로드) -> (DB 파일 메타데이터 등록 실패)
(DB 파일 메타데이터 등록) -> (물리적 파일 업로드 실패)
를 고려해봤습니다. 

뒤의 단계가 실패하면 앞의 단계도 롤백 해야한다 생각합니다. 

롤백 비용적 측면에서 DB 파일 메타데이터를 롤백 시키는 게 실제 서버에 올라간 데이터를 삭제하는 거 보다 저렴하다 생각해서 위와 같이 코드를 구성하였습니다.

pear96 replied:
오.. 그러네요 업로드 때 롤백을 고려해보면 DB 저장이 먼저 수행되는게 낫겠네요

VSFe commented on src/test/java/com/c4cometrue/mystorage/file/FileDataAccessServiceTest.java:
```diff
@@ -0,0 +1,66 @@
+
+import com.c4cometrue.mystorage.exception.ServiceException;
+
+@DisplayName("파일 데이터어세스 서비스 테스트")
```
국어 이슈

VSFe commented on src/main/java/com/c4cometrue/mystorage/exception/ServiceException.java:
```diff
@@ -0,0 +1,19 @@
+		this.code = errorCode;
+	}
+
+	public ServiceException(ErrorCode errorCode, Object... args) {
```
args의 역할이 여기서 무엇인가요? String이 Template을 형성하는 상태가 아닌 것 같아요.

VSFe commented on src/main/java/com/c4cometrue/mystorage/exception/ErrorCode.java:
```diff
@@ -0,0 +1,21 @@
+@AllArgsConstructor
+@Getter
+public enum ErrorCode {
+	UNAUTHORIZED_FILE_ACCESS(HttpStatus.UNAUTHORIZED, "비정상적인 요청입니다."),
```
로그인을 사용하는 상황이 아니다보니, (로그인을 했다고 쳐도, 다른 사람 파일에 접근하지 못하게 하는건 UNAUTHORIZED랑은 안 맞으니) HttpStatus.FORBIDDEN을 사용하는게 더 나은 것 같습니다.

kochungcheon replied:
넵 알겠습니다! 코멘트에 맞춰 수정했습니다

VSFe reviewed (changes requested):
![image](https://github.com/C4-ComeTrue/my-storage/assets/4595546/44252d31-8370-4942-b344-055e13533feb)

kochungcheon committed:
refactor: PR 반영

[FileDeleteRequest]
- Long -> long

[FileController]
- RESTFUL
- RequestMapping
- @ResponseStatus(HttpStatus.NO_CONTENT)

[FileService]
- findBy(fineId) -> findBy(fileId, userId)

[Metadata]
- validate 제거

[ServiceException]
- 오탈자 수정

[ServiceException]
- ServiceException(ErrorCode errorCode, Object... args) 제거

[ErrorCode]
- UNAUTHORIZED_FILE_ACCESS(HttpStatus.UNAUTHORIZED) ->UNAUTHORIZED_FILE_ACCESS(HttpStatus.FORBIDDEN)

[그 외]
- Service 레이어에 인자로 넘긴 DTO 분해
- 변경에 따른 테스트 코드 수정
- 사용하지 않을 수 있는 변수 제거함으로 테스트 코드 가독성 확보

kochungcheon committed:
refactor: PR 반영

[ErrorCode]
- FILE_DELETE_ERROR 추가
- DUPLICATE_FILE_NAME 추가

[FileDataAccessService]
- deleteBy에 existBy 추가
- persist에 duplicateBy 추가
- existBy 추가
- duplicateBy 추가

[FileReader]
- existBy 추가
- duplicateBy 추가

[FileRepository]
- checkDuplicateFileName 추가

[FileUtil]
- FileService 기술 관련 코드 추가

[Test]
- 변경 사항에 따른 테스트 코드 추가

kochungcheon committed:
refactor: PR 반영

[ServiceException]
- 양식 변경

[ErrorCode]
- 양식 변경

[Test]
- 변경 사항에 따른 테스트 코드 추가

VSFe commented on src/main/java/com/c4cometrue/mystorage/file/FileController.java:
```diff
@@ -13,26 +15,27 @@
-		return ResponseEntity.noContent().build();
+	@PostMapping
+	@ResponseStatus(HttpStatus.NO_CONTENT)
+	public void uploadFile(FileUploadRequest fileUploadRequest) {
```
`@ResponseStatus(HttpStatus.CREATED)`

kochungcheon replied:
어떤 반환 값을 줘야 할지 고민해보았습니다. 

서버에 데이터를 올린 것임으로 추천해주신 create 가 적절하다고 생각합니다

또한, 다른 컨트롤러들의 상태값도 검토해보았는 데요. 

현재는 어떤 내용도 반환하고 있지는 않지만, 특정 값을 반환할 일이 생기면 200으로 

바꾸는 것이 괜찮다 생각합니다.

VSFe commented on src/main/java/com/c4cometrue/mystorage/exception/ServiceException.java:
```diff
@@ -1,19 +1,45 @@
 import lombok.Getter;
 
 @Getter
 public class ServiceException extends RuntimeException {
```
Exception을 전체적으로 Handling 할 수 있는 수단이 필요합니다.
(지금은 예외를 반환한다고 절대 저희가 원하는 대로 결과가 나오지 않을거에요.)
`@ControllerAdvice` 등의 어노테이션을 활용하여 수정해주세요.

kochungcheon replied:
ServiceExceptionHandler 를 만들고 @RestControllerAdvice 통해 문제를 해결하였습니다.
또한, ErrorResponse 을 통해 반환 합니다!

디버그 메시지를 밖으로 노출하는 게 좋을 까? 라는 의문을 가지고 있습니다 
(개발 시에는 현행 유지하고, 운영 시에는 디버그 메시지 노출은 하지 않는 방향으로 가야하나 생각했습니다)

pear96 replied:
저는 에러 메세지만 반환하고, 디버그 메세지가 있을 경우 logger로 출력하면 되지 않을까 생각했습니다

kochungcheon replied:
확실히 그렇게 하는게 더 좋은거 같네요 do 에 디버그 반환하는거 내리겠습니다!

VSFe commented on src/main/java/com/c4cometrue/mystorage/file/FileDataAccessService.java:
```diff
@@ -13,16 +13,30 @@ public class FileDataAccessService implements FileReader, FileWriter {
-		);
+	public Metadata findBy(Long fileId, Long userId) {
+		return fileRepository.findByIdAndUploaderId(fileId, userId)
+			.orElseThrow(() -> ErrorCode.UNAUTHORIZED_FILE_ACCESS.serviceException("fileId : {}, userId : {}", fileId,
```
파일이 존재하지 않아도 UNAUTHORIZED_FILE_ACCESS가 뜨겠네요?
차라리 여기서는 UNAUTHORIZED_FILE_ACCESS를 뭉개고, FILE_NOT_FOUND로 퉁치는게 나을 것 같습니다.

kochungcheon replied:
넵! 하나로 퉁치겠습니다! 에러를 넘길 때, 더 적절한게 있나 여러 상황을 고려해보겠습니다.

VSFe commented on src/main/java/com/c4cometrue/mystorage/file/FileRepository.java:
```diff
@@ -1,6 +1,20 @@
 public interface FileRepository extends JpaRepository<Metadata, Long> {
+	Optional<Metadata> findByIdAndUploaderId(Long id, Long uploaderId);
+
+	@Query("SELECT CASE WHEN COUNT(m) > 0 "
```
해당 쿼리의 실행계획을 확인해주세요.

kochungcheon replied:
![image](https://github.com/C4-ComeTrue/my-storage/assets/18523603/aa40f6dc-eac6-45ea-acfe-158e832ec0ec)

확인 결과, 인덱스를 타지 않는 쿼리.. 였습니다! 수정하겠습니다!!

VSFe replied:
인덱스는 추가하면 돼죠 ㅎㅎ

kochungcheon replied:
![image](https://github.com/C4-ComeTrue/my-storage/assets/18523603/d080e74a-e1ff-4267-b478-4f0995881521)

넵 감사합니다! uploader_id 인덱스 추가하겠습니다!  uploader_id, original_file_name 를 인덱스로 추가한 결과 uploader_id 인덱스가 타지는 것을 확인했습니다!

kochungcheon replied:
인덱스를 테이블에 @Table(name = "Metadata", indexes = @Index(name = "index_uploaderId", columnList = "uploaderId"))를 추가해줬습니다.

쿼리 작성 하거나 JPA를 사용할 때는 항상 쿼리 실행 계획을 확인해보겠습니다!

VSFe replied:
좋습니다~
이후에 쿼리를 사용할 땐, 어떤 인덱스를 사용하는지 기록하시면 많은 도움이 될 거예요!

VSFe reviewed (commented):
쪼금만 더... 수정해봅시다!

(+ 수정사항을 반영했으면, 코멘트로 어떻게 바꿨는지/본인의 생각 등을 작성해주세요.)

kochungcheon committed:
refactor: PR 반영

[에러 메시지 관련]

- ErrorResponse 추가

- ServiceExceptionHandler 추가

[인덱스 관련]

- userId 에 인덱스 추가

[기타]

- 변경 사항 관련 테스트 코드 추가

kochungcheon committed:
refactor: 에러 ResponseDTO 수정

- 디버그 메시지 삭제

- 변화에 따른 테스트 코드 수정

VSFe reviewed (approved):
수고하셨습니다~
Step 2 부턴, 좀 더 빠르게 힘내봐요!

kochungcheon commented:
자세한 리뷰 달아주셔서 다들 감사합니다!
